New RMPath implementation

michaeltyson opened the pull request:
This is my RMPath implementation, which is cleaner, with revised geometry, direct access to the underlying CGPath for things like hit testing, and a bounding rect accessor.

With a bit of luck, this closes route-me/#64.

Also includes a fix to a display bug, and a crash fix.

michaeltyson committed:
Added gitignore

michaeltyson committed:
Merge branch 'master' of github.com:route-me/route-me

Conflicts:
	.gitignore

michaeltyson committed:
Fix crash when zooming

michaeltyson committed:
New RMPath implementation

- Cleaner implementation with revised geometry
- Direct access to path, for hit testing, etc
- Projected bounds accessor
- Fixed RMMapContents bug resulting in overlay corruption while zooming

michaeltyson commented:
SWEET!  Github can now automatically merge pull requests for us!  That's the best thing I've seen ever.

michaeltyson commented on MapView/Map/RMMapContents.m:
```diff
@@ -830,7 +836,7 @@
 
 -(float) zoom
 {
-        return [mercatorToTileProjection calculateZoomFromScale:[mercatorToScreenProjection metersPerPixel]];
```
Oh yes - This is unrelated, but I noticed that when retina mode is enabled, the reported zoom level is actually incorrect.  Now the reported zoom level is actually correct, as far as the displayed tiles are concerned.

vyskocil commented:
I tried the code in my application with a path composed with about 3000 points. It's perfect when 18 <= zoom <= 14 or 15, but with lower zoom levels the performance is very bad, it takes forever between each redraw, it's then much slower than the previous code. Did you also experience this ?

michaeltyson commented:
Oh dear, I hadn't noticed that!  I'll get back to it when I can, see if I can figure out why.

PS. eventually, we should try to reimplement this stuff like MapKit does, as CATiledLayers!

alarribeau commented:
We had a lot of issues with the old implementation while panning the map. Your new implementation works like a charm for us.

michaeltyson commented:
That's good to know!

I still haven't had time to address the performance issues Vlad mentioned, I'm afraid, but I'm hoping to get to it eventually.

michaeltyson committed:
RMPath performance enhancement: Only render path after drag/zoom motion ended

michaeltyson commented:
SHA: 57b14ff55ac453e6f2aa0e46c8bde214495f133b should fix the performance issues, and should hopefully now surpass the original RMPath significantly, as it waits for the end of a zoom/drag operation before doing the path render.  Much, much smoother.

sjhennion commented:
While I've definitely found the performance boost nice (and I can actually zoom in all the way with paths!), the redrawing is a little jarring.  Every time I pan the map the redraw reoccurs, and the path is skewed until I release the touch.  Is there any caching going on, even when no zoom changes are made?  In other words, if I reach a point on my map, the path is drawn, I pan away, and then pan back (never zooming in or out, which I could understand requiring a redraw), why is none the previous path draw not held onto?  

YMMV, but I'm going to stick to the older implementation for now.  I've made my path geometry much simpler and locked the zoom into levels that don't result in memory allocation issues.  This gives me very decent performance, and the path drawing is always the correct shape.  I can take screenshots if what I'm saying isn't understood.

japan2011 commented:
Is this New RMPath implementation on this route-me git(https://github.com/route-me/route-me/).

michaeltyson commented:
@sjhennion: Hmm, that definitely doesn't sound right!  What do you mean by skewed? Are you able to post a screenshot?
Part of the difficulty with avoiding redraw when panning is that it's (well, so I'm told) necessary to clip the path layer to the screen in order to avoid memory/drawing issues.  So, you pan, it clips, then draws, but when you pan again, it has to clip again, and consequently it has to redraw again.  If you pan when it's zoomed out enough so that it doesn't have to clip, it won't redraw.

But, that said, it shouldn't be skewing the path - sounds like something's going wrong.  I'd love to see a screenshot.

threeseed commented:
What is happening is as follows:

1) Draw a route with it zoomed far out e.g. across a country.

2) Zoom in to say city level. Everything is fine at this point.

3) As you pan the map the route line will load the tiles at the edges at which point it will redraw the route line. It gives the effect that the route line has been skewed and it is now fixing it.

michaeltyson commented:
Ahh, yeah, I'm with you - that's because I hold off redrawing the offscreen area of the path until movement stops, to keep it smooth. The non-redrawn parts of the path are just shown as repeated edge pixels.  But perhaps it could be mitigated by clipping not to the screen, but to an area a bit bigger than the screen.

You'd still get that effect when panning beyond the fixed area, but you'd see it less often.  Would that be acceptable?

threeseed commented:
It would be amazing if you could try and see.

I would be imagine that if the area is big enough to allow the user to do a decent swipe then it should be fine.

My hope is that bigger pans will result in tiles being loaded which amongst the flicker may not even notice the line changing too much.

Question:

You call recalculateGeometry for each addPointToXY. If we are adding lots of points (e.g. 1000) can we just comment this line out and do this recalculation at the end ? Any issues you would expect it to cause ?

michaeltyson commented:
I'll give it a shot and see how it performs.

That's a good point with calling recalculateGeometry for every point - It's probably worth adding a bulk drawing routine to add a large number of points:

`- (void)addPoints:(RMLatLong*)points count:(int)count;`

sjhennion commented:
http://imgur.com/a/GVtEX

The top image shows the skewing (perhaps stretching is a better term).  After seeing your offscreen drawing explanation I understand why it's happening, but as I said before when weighing the performance boost (which I'm working around) to the visual artifact, I'm going to to stick to the previous implementation for the time being.  Good progress none the less!

sjhennion commented:
Also, in regards to your clipping to an area greater than the screen size; in theory, there is a "maximum pan distance" a user can make, which is dependent upon the gesture length, gesture speed, and deceleration rate set in route-me.  With reasonable deceleration values, I'm pretty sure this is about one screen width/height unit.  That is, if the user places their finger all the way to the left and completes a pan gesture to the very right of the screen, under standard conditions the map would only pan over by the width of the screen.  I assume that if your clipping area was equal to (screensize)^3, with the screen set dead center as such:

::::::::::::
::::[_]::::
::::::::::::

where the :::: represent clipping area and the [_] is the screen, you would always have a buffer of 'one screen size' in any direction.

sjhennion commented:
Also, has anyone else tried this on a 3G with iOS 3.1.3?  While it works fine on the simulator and on my iPad 2, when running on my phone I get no paths drawn at all.  However, the old implementation works fine.

michaeltyson committed:
Merge branch 'master' of github.com:route-me/route-me

Conflicts:
	MapView/Map/RMPath.h
	MapView/Map/RMPath.m

michaeltyson committed:
Correctly cancel tile loader connection before retrying

Previously, when a tile load fails and retries, the old NSURLConnection is not cancelled, and the HTML body associated with the error is incorrectly interpreted and saved as tile data. This cancels the connection before retrying.

michaeltyson commented:
Cancelling this pull request - I don't have time to resolve all the raised issues on my RMPath implementation.

clarketus commented:
FYI: I just tried this on my iphone 3g running 3.1.3 and it works fine. There appear to be some rendering artifacts when you scroll the map down but overall it is very similar to my ipod touch running 4.3

Due to errors such as "CoreAnimation: failed to allocate 2228352 bytes." in my app, and as this branch seems to fix that issue and increase rendering speed, I might have a go at fixing up the remaining issues...

Can confirm the issue found in that screen shot above http://imgur.com/a/GVtEX on both devices

clarketus commented:
another FYI: 

reverting the commit 

https://github.com/michaeltyson/route-me/commit/57b14ff55ac453e6f2aa0e46c8bde214495f133b

seems to give similar performance to the old implementation, fix the memory issues I had as well as all the rendering artifacts that were introduced.
